Documentation/Proposals/NNNN-custom-test-execution-traits.md

```diff
@@ -0,0 +1,376 @@
+  @TaskLocal static var current: Self?
+}
+```
+
+Binding task local values requires using the scoped access
+[`TaskLocal.withValue()`](https://developer.apple.com/documentation/swift/tasklocal/withvalue(_:operation:isolation:file:line:)
+API though, and that would not be possible if `Trait` exposed separate methods
+like `setUp()` and `tearDown()`.
+
+For these reasons, I believe it's important to expose this trait capability
+using a single, scoped access-style API which accepts a closure. A simplified
+version of that idea might look like this:
+
+```swift
+public protocol Trait: Sendable {
+  // ...
+
+  // Simplified example, not the actual proposal
+  func executeTest(_ body: @Sendable () async throws -> Void) async throws
+}
+
+extension MockAPICredentialsTrait {
+  func executeTest(_ body: @Sendable () async throws -> Void) async throws {
+    let mockCredentials = APICredentials(apiKey: "...")
+    try await APICredentials.$current.withValue(mockCredentials) {
+      try await body()
+    }
+  }
+}
+```
+
+### Avoiding unnecessarily lengthy backtraces
+
+A scoped access-style API has some potential downsides. To apply this approach
+to a test function, the scoped call of a trait must wrap the invocation of that
+test function, and every _other_ trait applied to that same test which offers
+custom behavior _also_ must wrap the other traits' calls in a nesting fashion.
+To visualize this, imagine a test function with multiple traits:
+
+```swift
+@Test(.traitA, .traitB, .traitC)
+func exampleTest() {
+  // ...
+}
+```
+
+If all three of those traits customize test execution behavior, then each of
+them needs to wrap the call to the next one, and the last trait needs to wrap
+the invocation of the test, illustrated by the following:
+
+```
+TraitA.executeTest {
+  TraitB.executeTest {
+    TraitC.executeTest {
+      exampleTest()
+    }
+  }
+}
+```
+
+Tests may have an arbitrary number of traits applied to them, including those
+inherited from containing suite types. A naïve implementation in which _every_
+trait is given the opportunity to customize test behavior by calling its scoped
+access API might cause unnecessarily lengthy backtraces that make debugging the
+body of tests more difficult. Or worse: if the number of traits is great enough,
+it could cause a stack overflow.
+
+In practice, most traits probably will _not_ need to customize test behavior, so
+to mitigate these downsides it's important that there be some way to distinguish
+traits which customize test behavior. That way, the testing library can limit
+these scoped access calls to only the traits which require it.
+
+## Detailed design
+
+I propose the following new APIs:
+
+- A new protocol `CustomTestExecuting` with a single required `execute(...)`
+  method. This will be called to run a test, and allows the conforming type to
+  perform custom logic before or after.
+- A new property `customTestExecutor` on the `Trait` protocol whose type is an
```

Comment:
It occurs to me that types that conform to this protocol aren't actually responsible for executing the test, they're just adding before/after cladding to it. We might want to mull over the name a bit more. Thesaurus?

Reply:
Well they do work "surrounding" the test, but _also_ execute it.